dashboard/views.py

```diff
@@ -301,27 +301,43 @@ def grade_distribution(request, course_id=0):
     course_id = canvas_id_to_incremented_id(course_id)
 
     current_user = request.user.get_username()
-    grade_score_sql = "select current_grade,(select current_grade from user where sis_name=" \
-                      "%(current_user)s and course_id=%(course_id)s) as current_user_grade " \
-                      "from user where course_id=%(course_id)s and enrollment_type='StudentEnrollment';"
-    df = pd.read_sql(grade_score_sql, conn, params={"current_user": current_user,'course_id': course_id})
+
+    grade_score_sql = f"""select current_grade,
+       (select ab_test_course From course where id=%(course_id)s) as show_number_on_bars,
+    (select current_grade from user where sis_name=%(current_user)s and course_id=%(course_id)s) as current_user_grade
+        from user where course_id=%(course_id)s and enrollment_type='StudentEnrollment';
+                    """
+    df = pd.read_sql(grade_score_sql, conn, params={"current_user": current_user, 'course_id': course_id})
     if df.empty or df['current_grade'].isnull().all():
         return HttpResponse(json.dumps({}), content_type='application/json')
-    number_of_students = df.shape[0]
+
+    df['tot_students'] = df.shape[0]
     df = df[df['current_grade'].notnull()]
     df['current_grade'] = df['current_grade'].astype(float)
+    df['grade_avg'] = df['current_grade'].mean().round(2)
+    df['median_grade'] = df['current_grade'].median().round(2)
+    df['show_number_on_bars'] = df['show_number_on_bars'].apply(lambda x: True if x == 1 else False)
+
+    df.sort_values(by=['current_grade'], inplace=True)
+    df.reset_index(drop=True, inplace=True)
+    grades = df['current_grade'].values.tolist()
+    logger.debug(f"Grades distribution: {grades}")
+    binning_grade_value, index = find_binning_grade_value(grades)
+    df['current_grade'] = df['current_grade'].replace(df['current_grade'].head(index),
+                                                      binning_grade_value)
+
     if df[df['current_grade'] > 100.0].shape[0] > 0:
-        df['graph_upper_limit']=int((5 * round(float(df['current_grade'].max())/5)+5))
+        df['graph_upper_limit'] = int((5 * round(float(df['current_grade'].max()) / 5) + 5))
     else:
         df['current_grade'] = df['current_grade'].apply(lambda x: 99.99 if x == 100.00 else x)
-        df['graph_upper_limit']=100
-    average_grade = df['current_grade'].mean().round(2)
-    df['tot_students'] = number_of_students
-    df['grade_avg'] = average_grade
+        df['graph_upper_limit'] = 100
+
+    df['graph_lower_limit'] = df['current_grade'].min()
 
     # json for eventlog
     data = {
-        "course_id": course_id
+        "course_id": course_id,
+        "show_number_on_bars": df['show_number_on_bars'].values[0]
     }
     eventlog(request.user, EventLogTypes.EVENT_VIEW_GRADE_DISTRIBUTION.value, extra=data)
 
```

Comment:
I think we could simplify what data is passed to the front end, rather than passing the whole DataFrame. Looking in the browser on what's returned from this view, there are a number of columns that are always the same value. Maybe we could remodel the Response to use something like the following structure:

```
{
   "summary": {
      "current_user_grade": XX.X,
      "tot_students": XX,
      "show_number_on_bars": true,
      "grade_avg": XX.X,
      "median_grade": XX.X,
      "grade_upper_limit": XXX,
      "grade_lower_limit": XX.X
   },
   "grades": [XX.X, XX.X, ...]
}
```

Reply:
I like this one too... But just did not put the effort to do it

Reply:
I have created the issue https://github.com/tl-its-umich-edu/my-learning-analytics/issues/707. This can be taken up separetely